### PR TITLE
Rewrite Fedora build version detection

### PR DESCRIPTION
**Pull request by mihawk90**
**Changes**
Rewrite so we don't need to constantly update with every new Fedora
release. This is especially useful when Fedora and Jellyfin release
cycles don't line up.

Version selection is as follows:
* TARGET environment variable, which is currently used already
* Currently running Fedora version
* Hardcoded Fallback version that can be updated occasionally

**Issues**
related to #2351
companion PR to jellyfin/jellyfin#7614

### CODE REVIEW

**Comment by mihawk90** [comment]
Rebased onto v10.8.0

## Tests done
:heavy_check_mark: `make rpms`
:heavy_check_mark: `TARGET=fedora-34-x86_64 make rpms` (although EOL anyway)
:heavy_check_mark: `TARGET=fedora-35-x86_64 make rpms`
:heavy_check_mark: `TARGET=fedora-123 make rpms` (fails like it is supposed to)
:heavy_check_mark: `TARGET=gibberish make rpms` (fails like it is supposed to)

:x: `TARGET=epel-7-x86_64 make rpms`

This fails due to `Operation not permitted` on this:
https://github.com/jellyfin/jellyfin-web/blob/e727eed1d10aacb7052ebaada1bf5946368d08fe/fedora/jellyfin-web.spec#L33-L36

But from what I can tell this was introduced in #3246 on purpose, although I'm unclear in how this relates to the referenced issue #3222.

Either way, I hesitate to remove those lines for now seeing as they were intended to fix issues with the deployment. And since these packages are essentially just a collection of text files, even installing the Fedora Package on EPEL (although not exactly clean) should be fine.

:x: `TARGET=epel-8-x86_64 make rpms`
:x: `TARGET=rocky+epel-8-x86_64 make rpms`

These both fail on the same `npm ERR`:
```
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.ZhN4su
+ umask 022
+ cd /builddir/build/BUILD
+ cd jellyfin-web-10.8.0
+ npm ci --no-audit --unsafe-perm

> swiper@6.8.4 postinstall /builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/swiper
> node -e "try{require('./postinstall')}catch(e){}"

Love Swiper? Support Vladimir's work by donating or pledging: 
> On Patreon https://patreon.com/vladimirkharlampidi 
> On Open Collective https://opencollective.com/swiper

> core-js@3.20.2 postinstall /builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js:
> https://opencollective.com/core-js 
> https://patreon.com/zloirock 
> https://paypal.me/zloirock 
> bitcoin: bc1qlea7544qtsmj2rayg0lthvza9fau63ux0fstcz 

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)


> core-js-pure@3.19.1 postinstall /builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/core-js-pure
> node -e "try{require('./postinstall')}catch(e){}"


> jellyfin-web@10.8.0 prepare /builddir/build/BUILD/jellyfin-web-10.8.0
> node ./scripts/prepare.js

[webpack-cli] HookWebpackError: Not supported
    at makeWebpackError (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/HookWebpackError.js:48:9)
    at hooks.processAssets.callAsync.err (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:3059:12)
    at eval (eval at create (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:104:1)
    at err (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:510:26)
    at compilation.hooks.processAssets.tapAsync (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/copy-webpack-plugin/dist/index.js:708:13)
    at process._tickCallback (internal/process/next_tick.js:68:7)
-- inner error --
Error: Not supported
    at compilation.hooks.processAssets.tapAsync (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/copy-webpack-plugin/dist/index.js:704:13)
    at fn (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:509:9)
    at Hook.eval [as callAsync] (eval at create (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:102:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/tapable/lib/Hook.js:18:14)
    at cont (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:3056:34)
    at createChunkAssets.err (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:3104:10)
    at symbolIterator (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/neo-async/async.js:3485:9)
    at process._tickCallback (internal/process/next_tick.js:61:11)
caused by plugins in Compilation.hooks.processAssets
Error: Not supported
    at compilation.hooks.processAssets.tapAsync (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/copy-webpack-plugin/dist/index.js:704:13)
    at fn (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:509:9)
    at Hook.eval [as callAsync] (eval at create (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:102:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/tapable/lib/Hook.js:18:14)
    at cont (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:3056:34)
    at createChunkAssets.err (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:3104:10)
    at symbolIterator (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/neo-async/async.js:3485:9)
    at process._tickCallback (internal/process/next_tick.js:61:11)
child_process.js:669
    throw err;
    ^

Error: Command failed: webpack --config webpack.prod.js
    at checkExecSyncError (child_process.js:629:11)
    at execSync (child_process.js:666:13)
    at Object.<anonymous> (/builddir/build/BUILD/jellyfin-web-10.8.0/scripts/prepare.js:11:5)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)
    at startup (internal/bootstrap/node.js:283:19)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! jellyfin-web@10.8.0 prepare: `node ./scripts/prepare.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the jellyfin-web@10.8.0 prepare script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /builddir/.npm/_logs/2022-06-16T02_25_55_291Z-debug.log
error: Bad exit status from /var/tmp/rpm-tmp.ZhN4su (%build)
    Bad exit status from /var/tmp/rpm-tmp.ZhN4su (%build)
```

Unclear why this is or how to fix it.

**Comment by h1dden-da3m0n** [comment]
> node ./scripts/prepare.js
> 
> [webpack-cli] HookWebpackError: Not supported
> ...

Might be wrong, as I am not 100% familiar what exactly the prepare script does, but the last time I started the CI migration from ADO to GHA I set the EnvVar `SKIP_PREPARE=true` to presumably skip the prepare script when calling `npm ci --no-audit`. (the same can be found in the ADO workflows, so I assume 1 that the script is obsolete and 2 it might be disabled because it causes build errors :thinking: )

**Comment by mihawk90** [comment]
Seems like that doesn't produce any build at all (which the long comment in the prepare.js states as well, so not sure why disabling it even works).
But anyway, no luck.
```
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.Z9eYiw
+ umask 022
+ cd /builddir/build/BUILD
+ cd jellyfin-web-10.8.0
+ SKIP_PREPARE=true
+ npm ci --no-audit --unsafe-perm

> swiper@6.8.4 postinstall /builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/swiper
> node -e "try{require('./postinstall')}catch(e){}"

Love Swiper? Support Vladimir's work by donating or pledging: 
> On Patreon https://patreon.com/vladimirkharlampidi 
> On Open Collective https://opencollective.com/swiper

> core-js@3.20.2 postinstall /builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js:
> https://opencollective.com/core-js 
> https://patreon.com/zloirock 
> https://paypal.me/zloirock 
> bitcoin: bc1qlea7544qtsmj2rayg0lthvza9fau63ux0fstcz 

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)


> core-js-pure@3.19.1 postinstall /builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/core-js-pure
> node -e "try{require('./postinstall')}catch(e){}"


> jellyfin-web@10.8.0 prepare /builddir/build/BUILD/jellyfin-web-10.8.0
> node ./scripts/prepare.js

added 1762 packages in 20.727s
+ exit 0
Executing(%install): /bin/sh -e /var/tmp/rpm-tmp.VXPE7w
+ umask 022
+ cd /builddir/build/BUILD
+ '[' /builddir/build/BUILDROOT/jellyfin-web-10.8.0-1.el8.x86_64 '!=' / ']'
+ rm -rf /builddir/build/BUILDROOT/jellyfin-web-10.8.0-1.el8.x86_64
++ dirname /builddir/build/BUILDROOT/jellyfin-web-10.8.0-1.el8.x86_64
+ mkdir -p /builddir/build/BUILDROOT
+ mkdir /builddir/build/BUILDROOT/jellyfin-web-10.8.0-1.el8.x86_64
+ cd jellyfin-web-10.8.0
+ /usr/bin/mkdir -p /builddir/build/BUILDROOT/jellyfin-web-10.8.0-1.el8.x86_64/usr/lib64/jellyfin/jellyfin-web
+ /usr/bin/cp -r 'dist/*' /builddir/build/BUILDROOT/jellyfin-web-10.8.0-1.el8.x86_64/usr/lib64/jellyfin/jellyfin-web
/usr/bin/cp: cannot stat 'dist/*': No such file or directory
error: Bad exit status from /var/tmp/rpm-tmp.VXPE7w (%install)
```

I'm assuming the change is related to `npm` itself. RHEL 8 ships 6.14.11 while Fedora is on 8.3.1.

Searching for `npm ERR! code ELIFECYCLE` suggests it got something to do with a corrupted node_modules directory and cleaning it is supposed to help, but seeing as we're starting out without a node_modules directory to begin with, I'm not sure what I'm supposed to do here.

**Comment by h1dden-da3m0n** [comment]
my bad, overlooked the detail of the old NPM version, yeah that can cause problems.
However, there is another detail that I find odd in the spec file now.

This is how I got web to build in my earlier attempts to migrate from ADO to GHA:

https://github.com/jellyfin/jellyfin-web/blob/e4b8321ab254a2965aca85f2f5684ddc0e79db3e/.github/workflows/publish-container.yaml#L69-L77

This ended up producing the static JS & HTML files to throw into an NGINX container, so it should be what we are looking for here too. (obviously omit the `mv`)

**Comment by mihawk90** [comment]
Just tried that and it didn't work either... it doesn't seem to like the webpack in general:

```
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.Ent2jt
+ umask 022
+ cd /builddir/build/BUILD
+ cd /builddir/build/BUILD
+ rm -rf jellyfin-web-10.8.0
+ /usr/bin/gzip -dc /builddir/build/SOURCES/jellyfin-web-10.8.0.tar.gz
+ /usr/bin/tar -xof -
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ /usr/bin/gzip -dc /builddir/build/SOURCES/jellyfin-web-10.8.0.tar.gz
+ /usr/bin/tar -xof -
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd jellyfin-web-10.8.0
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ SKIP_PREPARE=true
+ npm ci --no-audit

> swiper@6.8.4 postinstall /builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/swiper
> node -e "try{require('./postinstall')}catch(e){}"

Love Swiper? Support Vladimir's work by donating or pledging: 
> On Patreon https://patreon.com/vladimirkharlampidi 
> On Open Collective https://opencollective.com/swiper

> core-js@3.20.2 postinstall /builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js:
> https://opencollective.com/core-js 
> https://patreon.com/zloirock 
> https://paypal.me/zloirock 
> bitcoin: bc1qlea7544qtsmj2rayg0lthvza9fau63ux0fstcz 

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)


> core-js-pure@3.19.1 postinstall /builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/core-js-pure
> node -e "try{require('./postinstall')}catch(e){}"


> jellyfin-web@10.8.0 prepare /builddir/build/BUILD/jellyfin-web-10.8.0
> node ./scripts/prepare.js

added 1762 packages in 23.823s
+ exit 0
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.kXozSv
+ umask 022
+ cd /builddir/build/BUILD
+ cd jellyfin-web-10.8.0
+ npm run build:production

> jellyfin-web@10.8.0 build:production /builddir/build/BUILD/jellyfin-web-10.8.0
> webpack --config webpack.prod.js

[webpack-cli] HookWebpackError: Not supported
    at makeWebpackError (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/HookWebpackError.js:48:9)
    at hooks.processAssets.callAsync.err (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:3059:12)
    at eval (eval at create (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:104:1)
    at err (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:510:26)
    at compilation.hooks.processAssets.tapAsync (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/copy-webpack-plugin/dist/index.js:708:13)
    at process._tickCallback (internal/process/next_tick.js:68:7)
-- inner error --
Error: Not supported
    at compilation.hooks.processAssets.tapAsync (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/copy-webpack-plugin/dist/index.js:704:13)
    at fn (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:509:9)
    at Hook.eval [as callAsync] (eval at create (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:102:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/tapable/lib/Hook.js:18:14)
    at cont (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:3056:34)
    at createChunkAssets.err (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:3104:10)
    at symbolIterator (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/neo-async/async.js:3485:9)
    at process._tickCallback (internal/process/next_tick.js:61:11)
caused by plugins in Compilation.hooks.processAssets
Error: Not supported
    at compilation.hooks.processAssets.tapAsync (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/copy-webpack-plugin/dist/index.js:704:13)
    at fn (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:509:9)
    at Hook.eval [as callAsync] (eval at create (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:102:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/tapable/lib/Hook.js:18:14)
    at cont (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:3056:34)
    at createChunkAssets.err (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/webpack/lib/Compilation.js:3104:10)
    at symbolIterator (/builddir/build/BUILD/jellyfin-web-10.8.0/node_modules/neo-async/async.js:3485:9)
    at process._tickCallback (internal/process/next_tick.js:61:11)
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! jellyfin-web@10.8.0 build:production: `webpack --config webpack.prod.js`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the jellyfin-web@10.8.0 build:production script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /builddir/.npm/_logs/2022-06-16T23_12_23_591Z-debug.log
error: Bad exit status from /var/tmp/rpm-tmp.kXozSv (%build)
```

At least we're not alone, Brian's COPR builds are failing as well:
https://copr.fedorainfracloud.org/coprs/brianjmurrell/jellyfin/build/4510579/

What throws me off is this:
```
[webpack-cli] HookWebpackError: Not supported
```
So I checked the webpack docs and it says
https://webpack.js.org/guides/getting-started/
> The minimum supported Node.js version to run webpack 5 is 10.13.0 (LTS)

So for sanity I checked both versions again.
Fedora:
```
z% node --version
v16.14.0
z% npm --version
8.3.1
```

RHEL8 (well, Alma, but still):
```
+ node --version
v10.24.0
+ npm --version
6.14.11
```

So this *should* be fine? What's up with "not supported" then, I don't get it.

**Comment by h1dden-da3m0n** [comment]
given this is NodeJs we talking about and usually we compile web with v14 and b4 that with v12 it may very well be that we are using APIs or fewatures introduced in versions of NodeJs **after** v10.
NodeJS v10 is very old and out of NodeJS support for quite a while already IIRC, yes I know RedHat handles life support, but still.

Given that I assume we wont have any luck with RHEL 8 unless we consider using Code Streams (IIRC that was what RedHat dubed the faster updating supported versions of packages).

**Comment by mihawk90** [comment]
Oh well, see I didn't even check when that version released, apparently 2018 :D And it went EOL in April last year.
Wondering why RedHat keeps that around when it's EOL and doesn't just offer multiple LTS versions if they don't wanna break people's setups that rely on 10.whatever.

But, no point thinking about it. At the end of the day it's your call if you want to merge regardless and just let RHEL be broken. After all, it's just a collection of web files so as mentioned earlier people *can* just install one of the Fedora packages on RHEL instead. It's not very clean, but I'm fairly certain it would work, since once it's all packed up, Node isn't required anymore right?

**Comment by h1dden-da3m0n** [comment]
RedHat being RedHat, keeping very old version alive due to their LTS support. That is the answer why they have such an old version that NodeJS may no longer support but RedHat will.

Regardless, they introduced [Application Streams](https://www.redhat.com/en/blog/introduction-appstreams-and-modules-red-hat-enterprise-linux) with REHL 8, which allow you to exactly fast track and install newer versions if you need them.

As long as our CI won't break I will try to advocate for the changes to get in to 10.8.1, however, I cannot make any promises as that is scheduled for later today.

**Comment by mihawk90** [comment]
> noticed we build against CentOS

I saw that before but I was confused by [Azure's OS info](https://dev.azure.com/jellyfin-project/jellyfin/_build/results?buildId=39210&view=logs&j=5404562b-245d-5392-77db-b338601c3067&t=7b7ec673-c1f7-47fc-8575-80ec83870f8b&l=5) in the `Initialize Job` step:
```
Ubuntu
20.04.4
LTS
```
Guessing this is just because it's Docker and the CentOS Docker is on CentOS 7.

CentOS 7 works just fine for the CI because of the `chown` you're doing on it (haven't looked into why that was introduced and doesn't really matter). Since I build in `mock`, chown isn't permitted so that's a local build issue.

TLDR it won't break the CI (unless you go CentOS 8, which might break it, I don't know).

**Comment by h1dden-da3m0n** [comment]
Yeah, ADO nor GHA offer any other Linux Distro runners than Ubuntu, to work around that we use a container to build, as you already spotted.

I will still simply validate on my local podman host, so we are extra sure CI wont fail during releasing :wink:

File: deployment/Dockerfile.fedora [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM fedora:33
+FROM fedora:36
```
Comment (h1dden-da3m0n) [inline deployment/Dockerfile.fedora:1]: Okay, good thing I tried playing through the workflow of our current CI.
It seems that `fedora:36` does not bundle `make`, nor will it get installed alongside the packages that we install below (odd that automake does not bundle it).
This obviously leads to a build error once we attempt to run the Makefile to build the fedora RPM. 
Regardless, if you add `make` to the `dnf install` call below it builds as you would expect.

Could you please add `make` to the package list below, so that it gets installed and won't error the CI, after that I consider this PR done from my point of view.

**Review by h1dden-da3m0n — CHANGES_REQUESTED** [review]

File: deployment/Dockerfile.fedora (hunk 1 above)
Comment (mihawk90) [reply deployment/Dockerfile.fedora:1]: > It seems that `fedora:36` does not bundle `make`, nor will it get installed alongside the packages that we install below (odd that automake does not bundle it).

That is a super odd choice by them since the Desktop versions ship it by default... but oh well. Added and pushed.

**Review by h1dden-da3m0n — APPROVED** [review]
